catalog/spatial/csw/spatial-csw-endpoint/src/main/java/org/codice/ddf/spatial/ogc/csw/catalog/endpoint/CswEndpoint.java

```diff
@@ -536,46 +554,103 @@ public TransactionResponseType transaction(CswTransactionRequest request) throws
     response.setVersion(CswConstants.VERSION_2_0_2);
 
     int numInserted = 0;
+    final Subject subject = SecurityUtils.getSubject();
     for (InsertAction insertAction : request.getInsertActions()) {
-      insertAction = transformInsertAction(insertAction);
-      CreateRequest createRequest = new CreateRequestImpl(insertAction.getRecords());
-      try {
-        CreateResponse createResponse = framework.create(createRequest);
-        if (request.isVerbose()) {
-          response.getInsertResult().add(getInsertResultFromResponse(createResponse));
-        }
+      final InsertAction transformInsertAction = transformInsertAction(insertAction);
+
+      List<Metacard> metacards = transformInsertAction.getRecords();
+      CompletionService<CreateResponse> completionService =
+          new ExecutorCompletionService<>(queryExecutor);
+
+      for (Metacard record : metacards) {
+        CreateRequest createRequest = new CreateRequestImpl(record);
+        Callable<CreateResponse> callable =
+            () -> {
+              try {
+                return framework.create(createRequest);
+              } catch (IngestException | SourceUnavailableException e) {
+                LOGGER.debug("Unable to insert record(s)", e);
```

Comment:
Do we need to log and throw here?

Reply:
We log the detailed exceptions to the log file and then throw a generic CswException that is returned to the caller.